### PR TITLE
overlay: lock staging directories

### PR DESCRIPTION
**Pull request by giuseppe**
lock any staging directory while it is being used so that another process cannot delete it.
    
Now the Cleanup() function deletes only the staging directories that are not locked by any other user.
    
Closes: https://github.com/containers/storage/issues/1915
    
Signed-off-by: Giuseppe Scrivano <gscrivan@redhat.com>

### CODE REVIEW

**Comment by edsantiago** [comment]
I vendored in your PR and ran podman tests. [tons of failures, all in seccomp](https://github.com/containers/podman/pull/22425/checks?check_run_id=24464599416), all timeouts. Too many failures to be coincidence, I think.

**Comment by giuseppe** [comment]
> I vendored in your PR and ran podman tests. [tons of failures, all in seccomp](https://github.com/containers/podman/pull/22425/checks?check_run_id=24464599416), all timeouts. Too many failures to be coincidence, I think.

sorry for wasting your time on this.

I've updated the PR and ran the Podman tests: https://github.com/containers/podman/pull/22573 which are green now

**Comment by edsantiago** [comment]
Thanks! podman + composefs test now in progress, https://github.com/containers/podman/pull/22425

**Comment by edsantiago** [comment]
Same thing: [timeouts in seccomp-policy](https://cirrus-ci.com/build/6002271308218368). Common factor seems to be podman-local (not remote) + sqlite (not boltdb) + fedora (not debian).

I also vendored c-common, so maybe the bug is there?

**Comment by edsantiago** [comment]
Confirmed.
```console
# cat /etc/containers/storage.conf 
[storage]
driver = "overlay"
runroot = "/run/containers/storage"
graphroot = "/var/lib/containers/storage"

[storage.options]
pull_options = {enable_partial_images = "true", use_hard_links = "false", ostree_repos="", convert_images = "true"}

[storage.options.overlay]
use_composefs = "true"

# go mod edit --replace github.com/containers/storage=github.com/giuseppe/storage@a5c63583a5f1ae8d9495716f2fd8f84755c64feb
...
# make
...

# bin/podman run --rm --seccomp-policy '' quay.io/libpod/alpine-with-seccomp:label ls
bin
dev
...
HANG!
```

**Comment by giuseppe** [comment]
ah no, this is another issue in the PR.  I've fixed it now.  It happens only when the registry rejects the range request, and I am going to debug now why quay does that with the `quay.io/libpod/alpine-with-seccomp:labels` image

**Review by mtrmac — CHANGES_REQUESTED** [review]
(Note to self: the history is https://github.com/containers/storage/pull/775#discussion_r569649992 , so this really needs to access other processes’ layers.)

The general intent of this does make sense to me; marking as “request changes” to make sure this is not merged prematurely.

**Comment by edsantiago** [comment]
Well, FWIW, with this and the c-i PR [podman tests now pass](https://github.com/containers/podman/pull/22425)

File: pkg/lockfile/lockfile.go
Comment (giuseppe) [reply pkg/lockfile/lockfile.go:447]: I've splitted the functions so that we don't call `UnlockFileEx` on Windows without having a lock.  Not sure how it is handled, as I've not tested it (nor I've a way to do it), but it seems incorrect to call unlock if we don't hold the lock.

File: drivers/overlay/overlay.go
Comment (giuseppe) [reply drivers/overlay/overlay.go:2174]: The staging directory returned now is the directory where the content for the layer will be stored.

A staging directory is something like `/var/lib/containers/storage/overlay/staging/12345/dir`, and the lock file is ``/var/lib/containers/storage/overlay/staging/12345/staging.lock`

A caller shouldn't care about anything except the directory where the files are expected to be written.

File: store.go
Comment (giuseppe) [reply store.go:2998]: it exists only to be used from the `containers-storage` tool.  `applydiff-using-staging-dir` was added to mirror `applydiff`, which applies the diff to an existing layer.

File: pkg/lockfile/lockfile.go
Comment (mtrmac) [reply pkg/lockfile/lockfile.go:447]: I asked for `unlockAndCloseHandle` in https://github.com/containers/storage/pull/1662#discussion_r1256114313 to make it clear to non-UNIX programmers (or programmers unfamiliar with this weird UNIX gotcha) that a non-unlocking close is impossible.

But then https://learn.microsoft.com/en-us/windows/win32/api/fileapi/nf-fileapi-lockfileex says that `CloseHandle` also unlocks locks…

---

WDYT about `unlockAndCloseHandle(explicitClose bool)`?

Ultimately I can live with the two functions split, as the PR is now. Maybe add a comment warning about the automatic unlock?

File: drivers/overlay/overlay.go
Comment (mtrmac) [reply drivers/overlay/overlay.go:2174]: My view is that the caller has no business reading any individual files either way; it essentially gets an ~opaque handle. (Now that the other APIs are being removed, `ApplyStagedLayerOptions` and `CleanupStagedLayer` both ask for a complete, presumably unmodified, `DriverWithDifferOutput`.)

That’s even more the case with ComposeFS, where the structure of the staged contents is no longer the traditional overlayFS filesystem.

Assuming the above way of thinking, I think it’s simpler and clearer for the overlay driver _itself_ to structure it so that the “handle” being passed is the top-level directory; affecting anything in parent directories is unusual and unexpected. “Unusual and unexpected” is not _automatically_ a blocker, but I think overlay is already complex and minimizing complexity is good. Or, at least, documenting the complexity, which is “the very least” thing I wish for.

File: store.go
Comment (mtrmac) [reply store.go]: Isn’t the primary purpose of `containers-storage` to run tests? (And possibly to inspect existing broken stores, for which read-only operations are sufficient.)

It’s a bit disappointing to maintain a code path for the rare use case; to have no test coverage for the atomic create+apply code path we actually _are_ going to exercise; and to pay for that with an extra lock/presence check on every `ApplyStagedLayer` call. Most of that is pre-existing, not the fault of this PR…

… but maybe we can leave `ApplyDiffFromStagingDirectory` around, then?

Or, well, change the `applydiff-using-staging-dir` and the calling test to create the layer atomically, and hope (?) that there are no external users.

**Comment by giuseppe** [comment]
pushed a new version that addresses all the comments.

**Comment by rhatdan** [comment]
@mtrmac waiting on you.

**Comment by giuseppe** [comment]
@mohanboddu should the jira label create automatically the issue?

**Comment by giuseppe** [comment]
I'd like to get this into the next release, can we move it forward?

File: pkg/lockfile/lockfile.go
Comment (mtrmac) [inline pkg/lockfile/lockfile.go:436]: The error / “lock taken” paths after this point don’t release the in-process lock

File: pkg/lockfile/lockfile_test.go
Comment (mtrmac) [inline pkg/lockfile/lockfile_test.go:218]: Actual tests of the “try” behavior (when locked, and unlocked) would be useful, for both RO and RW.

In new functions, not in a test named `TestLockfileName`,  nor changing what the existing `TestLockfileRead` does. (Dare I ask for 100% code coverage for these primitives?)

File: store.go
Comment (mtrmac) [inline store.go:1440]: If this is going to stay this way, please document the locking rules of the three relevant stores. (It’s non-obvious that the `rwLayerStore` and `roLayerStore` objects have different rules, and especially `containerStore` is ~invisible.)

File: store.go
Comment (mtrmac) [inline store.go:3004]: AFAICS `containerStore` does not need to be locked on the “layer already exists” path.

**Review by mtrmac — COMMENTED** [review]
Just to unblock progress, not a careful review I’m afraid.

File: store.go
Comment (mtrmac) [reply store.go]: I’d still prefer the tests to exercise the atomic creation code path, and not to add the test-only one here.

I’m not currently sure that’s blocking for me.

File: drivers/overlay/overlay.go
Comment (mtrmac) [reply drivers/overlay/overlay.go]: Still outstanding, somehow.

File: drivers/overlay/overlay.go
Comment (mtrmac) [reply drivers/overlay/overlay.go:2179]: … but there are also various error paths on callers. So maybe the staging area should only be unlocked when this succeeds?

Or maybe this does not matter?

Note to self: FIXME investigate.

**Comment by giuseppe** [comment]
> Just to unblock progress, not a careful review I’m afraid.

Thanks. Fixed the comments and pushed a new version

**Comment by edsantiago** [comment]
[Successful CI run](https://cirrus-ci.com/build/5495554052980736) with composefs in #22425, with latest (this morning's) main. Also a successful run this weekend but that was before the giant vendor merge. Either way, tentative LGTM from the passing-tests front.

File: store.go
Comment (mtrmac) [inline store.go:1441]: the read-only layer stores must _not_ be locked on entry.

**Comment by giuseppe** [comment]
@mtrmac thanks for the review, I've opened a PR: https://github.com/containers/storage/pull/1926

**Pull request merged**
